extensions/azdata/src/azdata.ts

```diff
@@ -134,21 +135,106 @@ export async function upgradeAzdata(outputChannel: vscode.OutputChannel): Promis
+		}); //update if available and user wants it.
+		return findAzdata(outputChannel); // now again find and return the currently installed azdata
+	} catch (err) {
+		// Don't block on this since we want the extension to finish activating without needing user input.
+		// Calls will be made to handle azdata not being installed
+		promptToInstallAzdata(outputChannel, userRequested).catch(e => console.log(`Unexpected error prompting to install azdata ${e}`));
+	}
+	return undefined;
+}
+
+/**
+ * Checks whether a newer version of azdata is available - and if it is then invokes the process of azdata upgrade.
  * @param currentAzdata The current version of azdata to check again
  * @param outputChannel Channel used to display diagnostic information
+ * @param userRequested true means that this operation by was requested by a user by executing an ads command.
  */
-export async function checkAndUpdateAzdata(currentAzdata: IAzdataTool, outputChannel: vscode.OutputChannel): Promise<void> {
+export async function checkAndUpgradeAzdata(currentAzdata: IAzdataTool, outputChannel: vscode.OutputChannel, userRequested: boolean = false): Promise<void> {
 	const newVersion = await discoverLatestAvailableAzdataVersion(outputChannel);
 	if (newVersion.compare(currentAzdata.version) === 1) {
-		const response = await vscode.window.showInformationMessage(loc.promptForAzdataUpgrade(newVersion.raw), loc.yes, loc.no);
-		if (response === loc.yes) {
-			await upgradeAzdata(outputChannel);
+		outputChannel.appendLine(loc.foundAzdataVersionToUpgradeTo(newVersion.raw, currentAzdata.version.raw));
+		promptToUpgradeAzdata(outputChannel, userRequested).catch(e => console.log(`Unexpected error prompting to upgrade azdata ${e}`));
+	}
+}
+
+async function promptToInstallAzdata(outputChannel: vscode.OutputChannel, userRequested: boolean = false): Promise<void> {
+	let response: string | undefined = loc.yes;
+	const config = <AzdataDeployOption>getConfig(azdataAutoInstallKey);
+	outputChannel.appendLine(loc.autoDeployConfig(azdataAutoInstallKey, config));
+	if (userRequested) {
+		outputChannel.appendLine(loc.userRequestedInstall);
+	}
+	if (config === AzdataDeployOption.never && !userRequested) {
+		outputChannel.appendLine(loc.skipInstall(config));
+		return;
+	}
+	if (config === AzdataDeployOption.userPrompt) {
+		response = await vscode.window.showErrorMessage(loc.couldNotFindAzdataWithPrompt, loc.yes, loc.no, loc.always, loc.never);
+		outputChannel.appendLine(loc.userResponseToInstallPrompt(response));
+	}
+	if (response === loc.always || response === loc.never) {
+		await setConfig(azdataAutoInstallKey, response);
+	}
+	if (response === loc.yes || response === loc.always) {
+		try {
+			await installAzdata(outputChannel);
+			vscode.window.showInformationMessage(loc.azdataInstalled);
+			outputChannel.appendLine(loc.azdataInstalled);
+		} catch (err) {
+			// Windows: 1602 is User cancelling installation/upgrade - not unexpected so don't display
```

Comment:
While we're doing this - this should be moved into the Win32 specific code instead of here.

Reply:
I'm also fine doing this in a separate PR - given the other stuff it's probably actually better for reviewing to not add additional stuff.

Reply:
ok postponing for later